### ISSUE TITLE
elm-test 0.19.0-beta4 sometimes hangs

### ISSUE THREAD

**Issue by brian-watkins** [rtfeldman/node-test-runner#274]
elm-test 0.19.0-beta4 hangs depending on whether certain functions are called by the tests or the module being tested. I know this sounds weird but here's what I mean:

Suppose you're testing the update function below:

```
type alias Model =
  { name : String }

type Msg
    = GetStuff (Result Http.Error String)

update : Msg -> Model -> (Model, Cmd Msg)
update msg model =
    ( model, Http.send GetStuff <| Http.getString "http://someFunWebsite.com" )
```

And here's the test:

```
testThings : Test
testThings =
    describe "something"
    [ test "it does something" <|
        \() ->
            App.update (App.GetStuff (Ok "Something")) { name = "Fun Person" }
                |> Tuple.first
                |> Expect.equal { name = "Cool Person" }
    ]
```

This test should fail, but when I run it with elm-test 0.19.0-beta4, the test is never executed and the elm-test process simply hangs with this output:

```
$ elm-test

Success! Compiled 2 modules.
Success! Compiled 3 modules.
```

The funny thing is that if you remove the `Http.send` command from the `update` function like so:

```
update : Msg -> Model -> (Model, Cmd Msg)
update msg model =
    ( model, Cmd.none )
```

then the elm-test process runs as expected and the test fails as expected.

I did some digging into this problem ... 

As you all know, the `subscriptions` function in the elm-test app (Test.Runner.Node.App) either returns `Sub.none` or a subscription to the `receive` port, depending on (among other things) whether a `Time.now` task (sent by the `run` function) has been processed. This all seems like it should work fine. 

However, I looked at the compiled JS (elmTestOutput.js) in both the good and the bad cases and noticed that in the bad case _Platform_dispatchEffects is dispatching effects in an order that causes the `receive` port not to be connected to receive incoming messages.

It looks like, each time _Platform_dispatchEffects runs, it gathers the subscriptions and commands and then iterates over the effect managers (which includes any ports), passing the subscriptions or commands to the appropriate manager. When `Test.Runner.Node.App` starts, _Platform_dispatchEffects is called and, because the `run` function returns a command, this triggers another call to _Platform_dispatchEffects to process this command. So there are two calls to _Platform_dispatchEffects upon startup of the `Test.Runner.Node.App` -- I'll call them (A) and (B).

Here's the order of relevant dispatched effects (upon start of the app) in the good case:

(A.1) The system calls subscriptions to figure out if there are any.
(A.2) The function returns `Sub.none`
(A.3) The `receive` port is set to not accept any messages
(A.4) The `Task` effect manager receives the `Time.now` command
(B.1) The `initOrUpdate` function is called with the current time value and updates the model
(B.2) The `subscriptions` function is called with the new model and the subscription to the `receive` port is returned
(B.3) The `receive` port is set to accept messages
(B.4) The `Task` effect manager receives no commands

In the bad case, we get the following order:

(A.1) The system calls subscriptions to figure out if there are any.
(A.2) The function returns `Sub.none`
(A.3) The `Task` effect manager receives the `Time.now` command
(B.1) The `initOrUpdate` function is called with the current time value and updates the model
(B.2) The `subscriptions` function is called with the new model and the subscription to the `receive` port is returned
(B.3) The `Task` effect manager receives no commands
(B.4) The `receive` port is set to accept messages
(A.4) The `receive` port is set to not accept any messages

So, in the bad case, after these two calls to _Platform_dispatchEffects, we're left with the `receive` port not being registered to receive messages -- it's as if `subscriptions` had returned `Sub.none`. Thus, when a message is sent to this port, it's never processed and the elm-test process just hangs. 

Why is the order different? If you look at the compiled JS, in the bad case you'll notice that the `Task` effect manager is registered *before* the `receive` port is registered. In the good case, the order is reversed. These are the relevant lines:

```
_Platform_effectManagers['Task'] = _Platform_createManager(elm$core$Task$init, elm$core$Task$onEffects, elm$core$Task$onSelfMsg, elm$core$Task$cmdMap);
```

```
var author$project$Test$Runner$Node$receive = _Platform_incomingPort('receive', elm$json$Json$Decode$value);
```

So, the order of these two lines seems to affect the order that _Platform_dispatchEffects iterates over the effect managers. 

I figure the difference in order is due to the fact that the elm compiler optimizes the compiled JS to only contain code that's actually used, which results in different compiled output.

So these are my observations -- I'm no expert on the Elm runtime so maybe I'm misinterpreting things, but it seems like a problem with the elm compiler is causing elm-test to hang. I guess this could be addressed by changing the way `Test.Runner.Node.App` starts up or by changing the way the compiler works. Maybe I need to open an issue on the compiler, but thought I'd post an issue here in case folks have a better idea what's going on. I'm on MacOS 10.13.5 with Node 10.9.0.

**Comment by rtfeldman** [comment]
Wow, great find and amazing writeup @brian-watkins!

I attempted a fix on `elm-test@0.19.0-beta5` - can you try it out and let me know if it fixes the problem for you locally?

**Comment by brian-watkins** [comment]
Yep! The change in elm-test@0.19.0-beta5 fixes the problem. Thanks for the quick turnaround!
